Mage/src/main/java/mage/players/PlayerImpl.java

```diff
@@ -4708,6 +4708,14 @@ public boolean moveCards(Set<? extends Card> cards, Zone toZone, Ability source,
                 List<ZoneChangeInfo> infoList = new ArrayList<>();
                 for (Card card : cards) {
                     fromZone = game.getState().getZone(card.getId());
+                    // 712.14a. If a spell or ability puts a transforming double-faced card onto the battlefield "transformed"
```

Comment:
Must put that code to `maybeRemoveFromSourceZone` here:

![shot_240405_194845](https://github.com/magefree/mage/assets/8344157/76d49843-f5b5-4b16-acdd-463f4e7593ba)

P.S. Looks like old code from a screenshot must be changed too

Reply:
Code like:
if transform param then
  check transformable and is permanent
else
  check is permanent

Reply:
Uhm should check that the back face of the card is permanent if transformable, or just the card?

Reply:
That should be good (running tests locally):
![image](https://github.com/magefree/mage/assets/34709007/1155a82b-d455-4bfe-a827-16032d71be82)

I leave the Nightbound TODO, not sure what that one is about, but the Life // Death one should be solved.